### PR TITLE
Add user choice between `U3` or `GPI2` as single-qubit native gate

### PR DESCRIPTION
**Pull request by Simone-Bordoni**
Make the user choose between U3 or GPI2 as single qubit native gate

Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Comment by Simone-Bordoni** [comment]
For now i am keeping both u3 decomposition and add the gpi2 decomposition. As the hard-coded gates decompositions are growing a lot i have decided to put them in a separate file.
Having a general u3 decomposition may be a good idea to implement future more genmeral decomposition by choosing all single qubit native gates. Let me know what you think, if you agree with this implementation i will proceed by hard-coding all gpi2 decomposition.
@stavros11 @BrunoLiegiBastonLiegi

File: src/qibo/transpiler/unroller.py [hunk 1]
```diff
@@ -1,120 +1,159 @@
...
 
     Args:
         circuit (:class:`qibo.models.circuit.Circuit`): circuit model to check.
+        native_gates (:class:`qibo.transpiler.abstract.NativeGates`):
+            native gates supported by the quantum hardware.
     """
     for gate in circuit.queue:
         if isinstance(gate, gates.M):
             continue
         if len(gate.qubits) == 1:
```
Comment (stavros11) [inline src/qibo/transpiler/unroller.py:62]: I think here you can simplify by handling the one- and two-qubit gates together. In the end you just need to check if 
```py
NativeGates.from_gate(gate) & native_gates
```
is non-empty. This check works in both cases.

It is also fine to raise the same error message. The gate name appears there so the user already knows if it's one or two qubit.

**Review by stavros11 — COMMENTED** [review]
Thanks @Simone-Bordoni, @BrunoLiegiBastonLiegi for looking into this. Generally, the mechanism of plugging native gates to the unroller looks good, except a few comments below. I also agree with moving the decompositions to a different file.

The only issue I see is that our current mechanism of decompositions and the way the unroller functions (`_translate_*` etc.) use the decompositions is not very scalable if we attempt to use them with different combinations of native gates and the code could also be simplified. Unfortunately I cannot think of a better proposal but if I come up with something or if you have any suggestions we can discuss. I have the impression that some kind of "inheritance" between decompositions could help, but I may be wrong.

File: src/qibo/transpiler/decompositions.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+    gates.Unitary,
+    lambda gate: [gates.U3(0, *u3_decomposition(gate.parameters[0]))],
+)
+u3_dec.add(
+    gates.FusedGate,
+    lambda gate: [gates.U3(0, *u3_decomposition(gate.matrix(backend)))],
+)
+
+# register the iSWAP decompositions
+iswap_dec = GateDecompositions()
```
[the remark this answers is not in the document]
Comment (stavros11) [reply src/qibo/transpiler/decompositions.py:158]: I think for two-qubits we are doing two passes, the first is converting the arbitrary two-qubit gate to arbitrary single-qubit + native two-qubit and the second converts the single-qubits to native.
In terms of code this happens here
https://github.com/qiboteam/qibo/blob/d71fac386019b60e3d2f7df6036830a468a26065/src/qibo/transpiler/unroller.py#L111
and the following lines.

That being said, I think we need a more strict mapping between the decompositions and what native gates each decomposition assumes. For example the `u3_dec` translates almost everything to U3 but still assumes that Z and RZ are natives. Similarly the `iswap_dec` still assumes that a bunch of single-qubit gates (that are not iSWAPs) are native gates. 
We need to come up with a better design where `u3_dec` translates everything to U3, but if the user passes `NativeGates.RZ | NativeGates.U3` to the `Unroller`, it should be smart enough to use both and get the optimal decompositions.

File: src/qibo/transpiler/abstract.py [hunk 3]
```diff
@@ -9,26 +9,57 @@
...
+    def default(cls):
+        """Return default native gates set."""
+        return cls.CZ | cls.GPI2 | cls.I | cls.Z | cls.RZ | cls.M
+
+    @classmethod
+    def from_gatelist(cls, gatelist: list):
+        """Create a NativeGates object containing all gates from a gatelist."""
+        natives = cls(0)
+        for gate in gatelist:
+            natives |= cls.from_gate(gate)
```
[the remark this answers is not in the document]
Comment (Simone-Bordoni) [reply src/qibo/transpiler/abstract.py:40]: This way you don't get an error if you create an empty NativeGates

**Comment by Simone-Bordoni** [comment]
> Thanks @Simone-Bordoni, @BrunoLiegiBastonLiegi for looking into this. Generally, the mechanism of plugging native gates to the unroller looks good, except a few comments below. I also agree with moving the decompositions to a different file.
> 
> The only issue I see is that our current mechanism of decompositions and the way the unroller functions (`_translate_*` etc.) use the decompositions is not very scalable if we attempt to use them with different combinations of native gates and the code could also be simplified. Unfortunately I cannot think of a better proposal but if I come up with something or if you have any suggestions we can discuss. I have the impression that some kind of "inheritance" between decompositions could help, but I may be wrong.

Maye tommorrow we should discuss better this point. I agree that scalability is an issue with this implementation but I don't have any ideas of how to solve the problem.

**Comment by Simone-Bordoni** [comment]
I have finisched correcting the "small issues" requested in the review. I think, once reviewed again, we should merge this PR in order to have a working decomposition with GPI2. In the future we can try to address the problem of general decompositions when we will have more ideas.

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]

**Pull request merged**